config/environments/test.rb

```diff
@@ -59,8 +59,6 @@
     Bullet.enable = true
     Bullet.bullet_logger = true
     Bullet.raise = true # raise an error if n+1 query occurs
-    Bullet.add_whitelist type: :unused_eager_loading, class_name: "Transaction", association: :provider
```

Comment:
I am wondering if this is actaully the 'correct' place to add whitelists as putting them in before and after hooks mean we keep adding the same thing over and over?

Reply:
I take your point. I suspect the 2 points may be connected. If the config is reapplied after each test, that could be why I'm able to overwride it in my own before block that's run later. I'll check how much overhead this move removes from to the test suite run time.

Reply:
Yes, with my version it looks like the Bullet whitelist hash is reset after every test. 

Moving these lines to test.rb does not appear to save any noticeable time to the running of the test suite, both at 1 minute 54 seconds ish on my machine. I guess this makes sense as it is building a small hash that doesn't do any further calculations.

Their documentation does not set out a convention for us to follow on where whitelists should be configured nor trade offs to consider. Moving whitelists that do not change to initialisation rather than `before(:each)` does seem reasonable. I think the trade off is we lose the ability to control these whitelists on a per test basis - something we need for our service classes in this PR.

If we move the whitelists to test.rb I think the only alternative is to turn Bullet off for the entire service class test before, and re-enable it afterwards. This also feels like a worse choice since we would have no monitoring at all.

I'd be in favour of keeping as much bullet monitoring as we can by leaving this in bullet.rb unless there's a way I've not seen to avoid doing `Before … bullet.enable = false`, `After … Bullet.enable = true`?

Reply:
@leeky I'm concious that @mec may be heads down in other work. Do you have an opinion on our best course of action here?

Reply:
@tahb here is a quick thing you can do to demonstate what I am seeing that I would lke to understand:

In `spec/support/bullet.rb` add a line after 13:

`pp Bullet.whitelist`

So we print out the current whitelist after each spec runs, now run the suite and watch that whitelist grow! 

Perhaps it doesn't matter that the whitelist is growing exponentially or I am misunderstanding, if so I would like to understand.

We are using the before hook for adding the `:parent` association on Activity in three spec files so I believe the hooks are scoped to the specs inside that file. That means we see three entries in the whitelist for `:parent` three times (three for each file). Also notice once the three specs have run the three `:parent` entries remain in the whitelist for the rest of the suite, which again points to the global nature of using `add_to_whitelist`.

Rpsec itself is configured to add four entries to the whitelist for every spec in the suite, so there a lot of entries in the whitelist. This is what makes me want to put the existing `Bullet.add_whitelist` in one place rather than in every spec.

Ultimately, this is how bullet has been setup so I am not trying to block this - I would just like to learn and understand, if we want to just move this forward, I think it is fine in a wider context!

Reply:
I can now see the whitelist growing linearly, though I'm not sure of the immediate impact it clearly isn't a healthy trait. 

I had wondered which set of problems we'd prefer: sensible whitelist configuration or making sure Bullet is monitoring for future performance issues in our classes. 

I've found another way which should give us both with `Bullet.delete_whitelist`. Unfortunantley I didn't spot this sooner!

```
  before(:all) do
    Bullet.add_whitelist type: :unused_eager_loading, class_name: "Activity", association: :parent
  end

  after(:all) do
    Bullet.delete_whitelist(type: :unused_eager_loading, class_name: "Activity", association: :parent)
  end
```